wip: Substepping

baccigalupi opened the pull request:
@ScottONeal This is a preview of how I was using substep implementation to rethink the kinds of modules that would make this code more reasonable. The way that logging, running steps and status was all mixed together was making it hard for me to take things apart for this substep stuff.

Instead of these mixed up objects, I was moving to value objects with events and listening. These are lightweight objects that just present the underlying data better for ease of understanding in other places. There are also collections for these value objects, value collections. Finally there are runner and collection runner classes. 

This work doesn't yet fully use the emitted status information to build out the ending status report.

As a next step, all those classes that are currently in one big file, I was going to break out and test. I am guessing that when I do that it will become clear that some responsibilities need to shuffle a bit. Once I have those extracted, I want to figure out how to insert them into the current StepCollection/Step world and then use events to write out the reports.

Let me know what you think. This is in progress, so don't merge or approve it yet!

Commit:
Substep API stubbed out in World

        * Add cucaroo tests to the success feature that call substeps
        * Makes the cucaroo tests not fail, but does not actually run
          substeps yet

Commit:
Builds out substeps implementation

        Creates a number of alternative classes that separate printing,
        formatting and status from running. Uses events to radiate
        activity up instead of passing everything down.

        This does not yet print results when things go wrong in a
        substep.

baccigalupi committed:
Extracts and tests simple step-value object

baccigalupi committed:
Extracts StepValueCollection from Substeps

baccigalupi committed:
Extracts MarrySteps into own module and refactors

ScottONeal commented on test/unit/step-value-test.js:
```diff
@@ -0,0 +1,52 @@
+'use strict';
+
+const assert = require('assert');
```
I've been using `chai expect` a bit and have been liking it. It can auto generate errors on failed asserts for something like: `expect(false).to.be.ok`

ScottONeal commented on test/unit/step-value-test.js:
```diff
@@ -0,0 +1,52 @@
+  it('makes available the most important parts of the step', function() {
+    assert.equal(stepValue.text, 'I am the step text');
+    assert.equal(stepValue.type, 'Given');
+    assert.equal(stepValue.implementation, matchingDefinitions[0].implementation);
```
You would probably want this to be a `deepEqual`, since this just asserts that the references are the same. Unless you are wanting to track you definition references, while passing them around.

ScottONeal commented on lib/step-value.js:
```diff
@@ -0,0 +1,29 @@
+const EventEmitter  = require('events').EventEmitter;
+
+class StepValue {
+  constructor(compiled, definitions) {
```
You may want to add an argument default to `definitions`, so that we can guarantee `definitions.length` will return a `number`

so, `constructor(compiled, definitions = [])`. You could default compiled as well, but maybe throw an error if not?

ScottONeal commented on lib/step-value-collection.js:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  listen(listener) {
+    this.collection.forEach((item) => {
```
Since `this` is nested below and can change depending on your caller. I would assign it, `let self = this`

baccigalupi replied:
I thought the es6 arrow functions handled this. I have been running more tests, and not having the usual deref issues that I experience with JS

ScottONeal replied:
You are correct, I forgot about this! https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

ScottONeal commented on lib/step-value-collection.js:
```diff
@@ -0,0 +1,43 @@
+    this.collection.forEach((item) => {
+      ['error', 'pending'].forEach((eventType) => {
+        let wrappedListener = () => {
+          this.emit(eventType, item);
```
`self.emit(eventType, item);`

ScottONeal commented on test/unit/step-value-collection-test.js:
```diff
@@ -0,0 +1,78 @@
+    collection = new StepValueCollection([validStep, ambiguousStep, ambiguousStep]);
+    let ambiguousSteps = collection.ambiguousSteps();
+    assert.equal(ambiguousSteps.length, 2);
+    assert.equal(ambiguousSteps[0], ambiguousStep);
```
Not sure if this is intended, but this assertion makes sure that the references are equal not the actual struct. As a consumer of `StepValueCollection` do I need to make sure that it keeps the original step references or do I not care if it modifies/copies/etc under the covers? If the ladder, then we should probably change these assertions to `deepEqual`

ScottONeal commented on test/unit/step-value-collection-test.js:
```diff
@@ -0,0 +1,78 @@
+
+    validStep.emit('error', 'hi');
+
+    assert.equal(stepListener.callCount, 2);
```
I'm sure I'm missing something. Why is the `callCount` 2 and not 1?

baccigalupi replied:
Well it is test setup. The `validStep` object is in the collection twice for test setup convenience. Should probably change for clarity.

ScottONeal commented on lib/substep.js:
```diff
@@ -0,0 +1,152 @@
+
+  runAll(callback) {
+    let steps = new MarrySteps(this.stepTexts, this.definitions.steps).marry();
+    steps.timeout = this.timeout;
```
Should we pass this timeout to MarrySteps some how, and it adds the `timeout` to the steps?

baccigalupi replied:
I am not sure how all the pieces go together yet. Timeout and also listening to step and collection for printing/status aggregation is still a bit of a mystery that I hope becomes clear. If I have to pass it down a ton of times, probably the wrong place :(

ScottONeal reviewed (commented):
I think this is a good direction to move towards. Eventually, it would be nice to have a `reporter` harness in which we could have different formats for output. ie: `CLI`, `html`, `json`. 

Prima facie, `SubSteps` as a class name was a little hard for me to associate it's purpose. However, I think I've grok'ed it now. 

Good 🧀 !

baccigalupi committed:
Extracts StepRunner from SubStep

        Also, switches to using EventEmitter2

baccigalupi committed:
Gets StepCollectionRunner running, with changes to StepRunner

baccigalupi committed:
Small refactor changes to what gets passed around

baccigalupi committed:
Tests and refactors a bit of the SubStep

baccigalupi committed:
StepRunner triggers events on StepValue before and after

baccigalupi committed:
Starts to build out Suite status and event driven printing

        * Duplicates StepPrinter to StepValuePrinter, but tailored to
          value object
        * Creates SuiteObserver with status contained therein
        * Tests

baccigalupi committed:
Changes Runner to SuiteRunner

baccigalupi committed:
Integrates suite runner in SubSteps

baccigalupi committed:
Complete change the world WIP commit

After substep running looked to be working. I took this big step
getting all the more normal flowing work to use events and an observer
instead of the intermingling of responsibilities in the old way.

  * Still to do, get tests passing (bad refactoring work).
  * Suite report prints at end instead of data blob
  * More cucaroo tests to assure that
    * nested substeps work
    * other types of failure in substeps work

baccigalupi committed:
WIP: not so working substep failures

baccigalupi committed:
Brings back the status reporting at the end

baccigalupi committed:
Substep failures handled crude but consistent with errors

baccigalupi committed:
Brings back the right exit status

ScottONeal committed:
adding "bin" param to config, bumping version

ScottONeal committed:
bumping version to 0.3.2

ScottONeal committed:
default featuresDirectory if config file is present but no featuresDirectory is present

ScottONeal committed:
removed accidently committed extra new line.

baccigalupi committed:
Merge pull request #9 from baccigalupi/default-features-directory-on-config

Default features directory on config

baccigalupi committed:
Merge branch 'master' into substepping

baccigalupi commented:
I merged this one in since we are already using it as the published version!
